### PR TITLE
SI-7899 Allow by-name inference outside of -Xstrict-inference

### PR DESCRIPTION
**Pull request by retronym**
As usual, the hole has been exploited in the wild. While you can't abstract
over by-name-ness without running into the ClassCastException or an un-applied
Function0, there are cases like the enclosed test where you can tiptoe around
those cases.

I'll try to change Scalaz to avoid using this (it's only used in one file) but
until then this commit conditionally moves the sound inference behind the
recently added -Xstrict-inference.

### CODE REVIEW

**Comment by retronym** [comment]
Review by @paulp, please.

/cc @gkossakowski this should be in 2.11.0-M6 if its not too late.

**Comment by paulp** [comment]
LGTM, but: while it's a good plan to begin corralling the instances of known unsoundness which can't immediately be fixed, -Xstrict-inference will need a lot of possibly tricky work before it is usable in practice (which can be seen trying to compile any nontrivial source with it.) It would be nice if it were an option which soundness-ambitious types could actually use all the time, or ever outside of test-case sized files.

File: test/files/run/t7899-regression.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  trait Monad[M[_]] {
+    def foo[A](ma: M[A])(f: M[A] => Any) = f(ma)
+  }
+  implicit def function1Covariant[T]: Monad[({type l[a] = (T => a)})#l] =
+    new Monad[({type l[a] = (T => a)})#l] {}
+
+  def main(args: Array[String]) {
+    // inference of T = (=> Any) here was outlawed by SI-7899 / 8ed7099
+    // but this pattern is used in Scalaz in just a few places and caused
+    // a regreesion.
```
Comment (paulp) [inline test/files/run/t7899-regression.scala]: The spelling of regression has regressed.

**Comment by retronym** [comment]
Typos fixed. I still have ambitions of rescuing this one from behind `-Xstrict-inference` before the next milestone, but I'll need to tweak Scalaz first. At that point, I'll probably add a `-Yallow-by-name-inference` in case other libraries need time to adjust.

**Comment by retronym** [comment]
PLS REBUILD ALL

```
     [get] Error opening connection java.io.IOException: Server returned HTTP response code: 504 for URL: http://private-repo.typesafe.com/typesafe/ivy-releases/org.scala-sbt/interface/0.12.4/jars/interface.jar
```

**Comment by retronym** [comment]
Closing for discussion. I have a patch against Scalaz that avoids the need for this, so maybe I'll resubmit the option to infer by-names as a `-Y` (as an insurance for other projects.)

**Pull request closed without merging**